lib/cosmos/io/json_drb.rb

```diff
@@ -100,57 +112,76 @@ def graceful_kill
+
+          # Create an http server to accept requests from clients
           begin
-            while true
-              begin
-                socket = @listen_socket.accept_nonblock
-              rescue Errno::EAGAIN, Errno::ECONNABORTED, Errno::EINTR, Errno::EWOULDBLOCK
-                read_ready, _ = IO.select([@listen_socket, @thread_reader])
-                if read_ready and read_ready.include?(@thread_reader)
-                  # Thread should be killed
-                  break
-                else
-                  retry
-                end
+            server_config = {
+              :Host   => hostname,
+              :Port   => port,
+              :Silent => true,
+            }
+
+            # The run call will block until the server is stopped.
+            Rack::Handler::Puma.run(JsonDrbRack.new(self), server_config) do |server|
+              @server_mutex.synchronize do
+                @server = server
               end
+            end
 
-              if @acl and !@acl.allow_socket?(socket)
-                Cosmos.close_socket(socket)
-                next
-              end
-              # Create new thread for new connection
-              create_client_thread(socket)
+            # Wait for all puma threads to stop before trying to close
+            # the sockets
+            start_time = Time.now
+            while true
+              puma_threads = false
+              Thread.list.each {|thread| puma_threads = true if thread.inspect.match(/puma/)}
+              break if !puma_threads
+              break if (Time.now - start_time) > 10.0
+              sleep 0.25
             end
-          rescue Exception => error
-            Logger.error "JsonDRb listen thread unexpectedly died.\n#{error.formatted}"
+
+            # Puma doesn't clean up it's own sockets after shutting down,
+            # so we'll do that here.
+            @server_mutex.synchronize do
+              @server.binder.close()
+            end
+
+          # The address in use error is pretty typical if an existing
+          # CmdTlmServer is running so explicitly rescue this
+          rescue Errno::EADDRINUSE
+            @server = nil
+            raise "Error binding to port #{port}.\n" +
+                  "Either another application is using this port\n" +
+                  "or the operating system is being slow cleaning up.\n" +
+                  "Make sure all sockets/streams are closed in all applications,\n" +
+                  "wait 1 minute and try again."
+          # Something else went wrong which is fatal
+          rescue => error
+            @server = nil
+            Logger.error "JsonDRb http server could not be started or unexpectedly died.\n#{error.formatted}"
             Cosmos.handle_fatal_exception(error)
           end
         end
+
+        # Wait for the server to be started in the thread before returning.
+        start_time = Time.now
+        while ((Time.now - start_time) < 5.0) and !server_started
```

Comment:
What happens if the server isn't started after 5s? Should an exception be raised?

Reply:
Added an exception and updated the spec to check it.